### PR TITLE
Added logic to convert the client data to XDM formatted data

### PR DESCRIPTION
**Pull request by addb**
## Motivation and Context

- Added MediaXDMEventHelper to generate XDM format data
- Added unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                }
+            }
+
+            return customMetadataList.sortedBy { it.name }
+        }
+
+        @JvmStatic
+        fun generateAdvertisingPodDetails(adBreakInfo: AdBreakInfo?): XDMAdvertisingPodDetails? {
+            val adBreakInfo = adBreakInfo
+            if (adBreakInfo !is AdBreakInfo) {
```
Comment (kevinlind) [inline code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt:134]: This technically works, but you are checking the type of the object in order to smart cast it to a non-optional. Instead, I'd suggest using the safe call operator `?.` to unwrap the object. You can do this in all places where are casting the object using `!is`

```
adBreakInfo?.let {
   return XDMAdvertisingPodDetails(it.name, it.position, it.startTime.toLong())
}

return null
```

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            MediaConstants.AdMetadataKeys.PLACEMENT_ID,
+            MediaConstants.AdMetadataKeys.SITE_ID
+        )
+
+        @JvmStatic
+        fun generateSessionDetails(mediaInfo: MediaInfo, metadata: Map<String, String>, forceResume: Boolean = false): XDMSessionDetails {
+            val sessionDetails = XDMSessionDetails()
+            sessionDetails.name = mediaInfo.id
+            sessionDetails.friendlyName = mediaInfo.name
+            sessionDetails.length = mediaInfo.length.toLong()
```
Comment (kevinlind) [inline code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt:71]: Calling `Double.toLong()` will round any fractional part down to zero, thus losing any fractional length. However, from what I can tell the XDM schema defines SessionDetails media content length as an Integer, not Float. If it is common for customers to use fractional values for "length", then we should consider changing the XDM schemas to use Floats instead of Integer and then update SessionDetails to use Double. This appears to be a common pattern for the other classes where lengths of type Double are converted to Long to conform to the XDM schemas.

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+            return XDMQoeDataDetails(
+                qoeInfo.bitrate.toLong(),
+                qoeInfo.droppedFrames.toLong(),
+                qoeInfo.fps.toLong(),
+                qoeInfo.startupTime.toLong()
+            )
+        }
+
+        @JvmStatic
+        fun generateErrorDetails(errorID: String): XDMErrorDetails? {
```
Comment (kevinlind) [inline code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt:242]: The return type can be made non-optional.

File: code/edgemedia/src/test/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelperTests.kt [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+    private val mediaInfo = MediaInfo.create("id", "name", "vod", MediaType.Video, 10.0)
+    private val adBreakInfo = AdBreakInfo.create("name", 1, 2.3)
+    private val adInfo = AdInfo.create("id", "name", 1, 10.0)
+    private val chapterInfo = ChapterInfo.create("name", 1, 2.3, 10.0)
+    private val qoeInfo = QoEInfo.create(1.1, 2.2, 3.3, 4.4)
+    private val muteStateInfo = StateInfo.create(MediaConstants.PlayerState.MUTE)
+    private val testStateInfo = StateInfo.create("testStateName")
+
+    val metadata = mutableMapOf("key1" to "value1", "key2" to "value2")
```
Comment (kevinlind) [inline code/edgemedia/src/test/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelperTests.kt:71]: For this custom metadata test data, place the values in a random order so the tests also verify the metadata is sorted.

File: code/edgemedia/src/test/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelperTests.kt
Comment (kevinlind) [inline code/edgemedia/src/test/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelperTests.kt:1]: For each MediaXDMEventHelper function which returns an optional value, add tests which validate a null value is returned.

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt (hunk 2 above)
Comment (addb) [reply code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt:71]: This should be fine, we have double/float values in our public APIs in order to match the APIs for our previous SDKs.

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt [hunk 5]
```diff
@@ -74,39 +74,37 @@ internal class MediaXDMEventHelper {
...
-
-                        // Audio standard metadata cases
-                        MediaConstants.AudioMetadataKeys.ALBUM -> sessionDetails.album = value
-                        MediaConstants.AudioMetadataKeys.ARTIST -> sessionDetails.artist = value
-                        MediaConstants.AudioMetadataKeys.AUTHOR -> sessionDetails.author = value
-                        MediaConstants.AudioMetadataKeys.LABEL -> sessionDetails.label = value
-                        MediaConstants.AudioMetadataKeys.PUBLISHER -> sessionDetails.publisher = value
-                        MediaConstants.AudioMetadataKeys.STATION -> sessionDetails.station = value
-                    }
+                if (!standardMediaMetadataSet.contains(key)) {
```
Comment (kevinlind) [inline code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt:77]: Since `when` will only assign the value for the keys defined in the `when` block, is this check needed?

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt [hunk 6]
```diff
@@ -159,19 +149,29 @@ internal class MediaXDMEventHelper {
...
-
-                    when (key) {
-                        MediaConstants.AdMetadataKeys.ADVERTISER -> advertisingDetails.advertiser = value
-                        MediaConstants.AdMetadataKeys.CAMPAIGN_ID -> advertisingDetails.campaignID = value
-                        MediaConstants.AdMetadataKeys.CREATIVE_ID -> advertisingDetails.creativeID = value
-                        MediaConstants.AdMetadataKeys.CREATIVE_URL -> advertisingDetails.creativeURL = value
-                        MediaConstants.AdMetadataKeys.PLACEMENT_ID -> advertisingDetails.placementID = value
-                        MediaConstants.AdMetadataKeys.SITE_ID -> advertisingDetails.siteID = value
-                    }
+                if (!standardAdMetadataSet.contains(key)) {
```
Comment (kevinlind) [inline code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt:152]: Same as previous comment, as `when` only assigns the value for the defined keys, is this check needed?

**Review by kevinlind — APPROVED** [review]
